File: app/src/main/java/com/worldventures/dreamtrips/wallet/analytics/tokenization/TokenizationAnalyticsLocationCommand.java
package com.worldventures.dreamtrips.wallet.analytics.tokenization;

import android.support.annotation.Nullable;

import com.worldventures.dreamtrips.wallet.analytics.WalletAnalyticsCommand;
import com.worldventures.dreamtrips.wallet.domain.entity.lostcard.WalletCoordinates;
import com.worldventures.dreamtrips.wallet.domain.entity.lostcard.WalletLocation;
import com.worldventures.dreamtrips.wallet.service.lostcard.LostCardRepository;

import java.util.List;

import javax.inject.Inject;

import io.techery.janet.command.annotations.CommandAction;

@CommandAction
public class TokenizationAnalyticsLocationCommand extends WalletAnalyticsCommand {

   @Inject LostCardRepository lostCardRepository;

   private final TokenizationCardAction tokenizationCardAction;

   public TokenizationAnalyticsLocationCommand(TokenizationCardAction tokenizationCardAction) {
      super(tokenizationCardAction);
      this.tokenizationCardAction = tokenizationCardAction;
   }

   @Override
   protected void run(CommandCallback<Void> callback) throws Throwable {
      tokenizationCardAction.setCoordinates(getLastKnownCoordinates());
      super.run(callback);
   }

   @Nullable
   private WalletCoordinates getLastKnownCoordinates() {
      List<WalletLocation> locations = lostCardRepository.getWalletLocations();
      WalletLocation lastKnownLocation = (locations == null || locations.isEmpty()) ?
            null : locations.get(locations.size() - 1);
      return (lastKnownLocation == null) ? null : lastKnownLocation.coordinates();
   }

}

File: wallet/src/main/java/com/worldventures/wallet/analytics/tokenization/TokenizationAnalyticsLocationCommand.kt
package com.worldventures.wallet.analytics.tokenization

import com.worldventures.wallet.analytics.WalletAnalyticsCommand
import com.worldventures.wallet.domain.entity.lostcard.WalletCoordinates
import com.worldventures.wallet.service.lostcard.LostCardRepository
import com.worldventures.wallet.util.WalletLocationsUtil
import io.techery.janet.command.annotations.CommandAction
import javax.inject.Inject

@CommandAction
class TokenizationAnalyticsLocationCommand(private val tokenizationCardAction: TokenizationCardAction) : WalletAnalyticsCommand(tokenizationCardAction) {

   @Inject internal lateinit var lostCardRepository: LostCardRepository

   @Throws(Throwable::class)
   override fun run(callback: CommandCallback<Void>) {
      tokenizationCardAction.setCoordinates(fetchLastKnownCoordinates())
      super.run(callback)
   }

   private fun fetchLastKnownCoordinates(): WalletCoordinates? {
      val locations = lostCardRepository.walletLocations
      val lastKnownLocation = WalletLocationsUtil.getLatestLocation(locations)
      return lastKnownLocation?.coordinates
   }

}